operator/README.md

```diff
@@ -0,0 +1,173 @@
+| registrarImage | string | [CSI node driver registrar](https://github.com/kubernetes-csi/node-driver-registrar) docker image name | `quay.io/k8scsi/csi-node-driver-registrar:v1.1.0` |
+| pullPolicy | string | Docker image pull policy. either one of `Always`, `Never`, `IfNotPresent` | `IfNotPresent` |
+| logLevel | integer | PMEM-CSI driver logging level | 3 |
+| deviceMode | string | Device management mode to use. Supports one of `lvm` or `direct` | `lvm`
+| controllerResources | [ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | Describes the compute resource requirements for controller pod |
+| nodeResources | [ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | Describes the compute resource requirements for the pods running on node(s) |
+| nodeSelector | string map | [Labels to use for selecting Nodes](../README.md#run-pmem-csi-on-kubernetes) on which PMEM-CSI driver should run. | `{ "storage": "pmem" }`
+
+#### DeploymentStatus
+
+A PMEM-CSI Deployment's `status` field is a `DeploymentStatus` object, which has
+a `phase` field. The phase of a Deployment is high-level summary of where the
+Deployment is in it's lifecycle.
+
+The possible `phase` values and their meaning are as below:
+
+| value | meaning |
+|---|---|---|
+| Pending | The `Deployment` has been accepted by the system, but waiting for one or more of it's pre-requisites to be completed. Such as waiting for required [TLS certificates](../README.md#security) to be generated |
+| Initializing | The required sub-resources of the `Deployment` are being initialized |
+| Running | All the required sub-resources of the `Deployment` are created |
+| Failed | For some reason the state of the `Deployment` failed and cannot be progressed |
+
+### TLS Certificate approval
+
+All PMEM-CSI driver communication is protected by mutual TLS. So the driver
+deployment needs to generate those certificates. The operator uses Certificate
+Authority(CA) provided by the Kubernetes cluster to get those certificates. On a
+new deployment, the operator creates Kubernetes `CertificateSigningRequest`
+objects and keeps the deployment in `pending` state till those CSRs get approved
+by an authorized user so that Kubernetes CA could generate the certificates.
+Once the CSRs are approved and certificates are ready, then the deployment
+process gets progressed.
+
+If the operator is configured with appropriate RBAC rules with enough
+permissions to approve CSRs, the operator itself runs a dedicated certificate
+approval control-loop, that detects and approves the certificate signing
+requests generated by a PMEM-CSI Deployment. For validating if a CSR is
+generated by the operator it uses the `Username` field of
+`CertificateSigningRequestSpec`.
```

Comment:
And that `Username` cannot be faked by someone else?

Reply:
> And that `Username` cannot be faked by someone else?

Oops, this was talking about the initial idea, and no more valid now. That means we are not using `Username` check anymore, Just checking if the `CSR` is owned by a `Deployment`.

I will fix the documentation.

Reply:
Same question applies: can some rogue deployment trick PMEM-CSI into approving a CSR that it shouldn't approve?

Reply:
I doubt if some can create CSR object and assign the ownership to a running deployment object. Though I am not sure.

Reply:
Can you explain to me what you mean with ownership? I probably lack some background here.

Reply:
The [deployment controller adds](https://github.com/intel/pmem-csi/pull/525/commits/0fecdcdc055664608ab9f7fa16ae75a57c74d4e0#diff-ab6d9094a1503b34d88bc9ba1066c814R161-R172)  a [ower reference](https://kubernetes.github.io/cluster-registry/reference/build/index.html#ownerreference-v1) to all the objects it creates for a deployment. Certificate controller checks if a CSR is owned by a Deployment before approving it.

Reply:
Anyone can set arbitrary owner references in the objects that they are creating. How would the apiserver decide who's allowed to do that?

So this check doesn't provide any additional security. What may work is to restrict signing to CSRs in the namespaces managed by the operator. Then we can argue that an adversary first would have to gain the privilege to create CSRs in that namespace, something that a non-admin users shouldn't have. Once an adversary has that privilege, he probably can also meddle with the other entities.

Reply:
I changed the CSR approver, now it considers that:
  - If the CSR is owned by a valid Deployment, and that deployment is in _Pending_ state
  - the CSR subject is one of the valid common names expected by PMEM-CSI driver.